File: Medium/2025-08-23-3192-FindTheMinimumAreaToCoverAllOnesII/solution.py
from typing import List

class Solution:
    
    def boundingBoxArea(self, grid: List[List[int]], u: int, d: int, l: int, r: int) -> int:
        min_row, max_row = len(grid), -1
        min_col, max_col = len(grid[0]), -1

        for i in range(u, d + 1):
            for j in range(l, r + 1):
                if grid[i][j] == 1:
                    min_row = min(min_row, i)
                    max_row = max(max_row, i)
                    min_col = min(min_col, j)
                    max_col = max(max_col, j)
        
        # If no 1's is found, return "infinite" cost so this partition is invalid
        if max_row == -1:
            return float('inf')
        
        return (max_row - min_row + 1) * (max_col - min_col + 1)

    def solve(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        best = n * m
        
        # Case 1: L- shaped partitions
        for row in range(n - 1):
            for col in range(m - 1):
                # Top + Bottom - left + Bottom - right
                area1 = (self.boundingBoxArea(grid, 0, row, 0, m - 1) + 
                         self.boundingBoxArea(grid, row + 1, n - 1, 0, col) +
                         self.boundingBoxArea(grid, row + 1, n - 1, col + 1, m - 1))

                # top-left + top-right + bottom
                area2 = (self.boundingBoxArea(grid, 0, row, 0, col) + 
                         self.boundingBoxArea(grid, 0, row, col + 1, m - 1) +
                         self.boundingBoxArea(grid, row + 1, n - 1, 0, m - 1))

                best = min(best, min(area1, area2))

        # Case 2: Three horizontal strips
        for r1 in range(n - 2):
            for r2 in range(r1 + 1, n - 1):
                area = (self.boundingBoxArea(grid, 0, r1, 0, m - 1) + 
                        self.boundingBoxArea(grid, r1 + 1, r2, 0, m - 1) +
                        self.boundingBoxArea(grid, r2 + 1, n - 1, 0, m - 1))
                best = min(best, area)
        
        return best

    def rotate90(self, grid: List[List[int]]) -> List[List[int]]:
        n, m = len(grid), len(grid[0])
        rotated = [[0] * n for _ in range(m)]

        for i in range(n):
            for j in range(m):
                rotated[j][n - 1 - i] = grid[i][j]
        
        return rotated

    def minimumSum(self, grid: List[List[int]]) -> int:
        rotated = self.rotate90(grid)
        return min(self.solve(grid), self.solve(rotated))

File: Medium/2025-08-23-3192-FindTheMinimumAreaToCoverAllOnesII/solution.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minimumSum = function(grid) {
    
    function boundingBoxArea(grid, u, d, l, r) {
        let minRow = grid.length, maxRow = -1;
        let minCol = grid[0].length, maxCol = -1;

        for (let i = u; i <= d; i++) {
            for (let j = l; j <= r; j++) {
                if (grid[i][j] === 1) {
                    minRow = Math.min(minRow, i);
                    maxRow = Math.max(maxRow, i);
                    minCol = Math.min(minCol, j);
                    maxCol = Math.max(maxCol, j);
                }
            }
        }
        // If no 1's is found, return "infinite" cost so this partition is invalid
        if (maxRow === -1) {
            return Number.MAX_SAFE_INTEGER / 3;
        }

        return (maxRow - minRow + 1) * (maxCol - minCol + 1);
    }

    function solve(grid) {
        const n = grid.length;
        const m = grid[0].length;
        let best = n * m;
        
        // Case 1: L- shaped partitions
        for (let row = 0; row + 1 < n; row++) {
            for (let col = 0; col + 1 < m; col++) {
                // Top + Bottom - left + Bottom - right
                const area1 = boundingBoxArea(grid, 0, row, 0, m - 1) + 
                             boundingBoxArea(grid, row + 1, n - 1, 0, col) +
                             boundingBoxArea(grid, row + 1, n - 1, col + 1, m - 1);

                // top-left + top-right + bottom
                const area2 = boundingBoxArea(grid, 0, row, 0, col) + 
                             boundingBoxArea(grid, 0, row, col + 1, m - 1) +
                             boundingBoxArea(grid, row + 1, n - 1, 0, m - 1);

                best = Math.min(best, Math.min(area1, area2));
            }
        }

        // Case 2: Three horizontal strips
        for (let r1 = 0; r1 + 2 < n; r1++) {
            for (let r2 = r1 + 1; r2 + 1 < n; r2++) {
                const area = boundingBoxArea(grid, 0, r1, 0, m - 1) + 
                            boundingBoxArea(grid, r1 + 1, r2, 0, m - 1) +
                            boundingBoxArea(grid, r2 + 1, n - 1, 0, m - 1);
                best = Math.min(best, area);
            }
        }
        return best;
    }

    function rotate90(grid) {
        const n = grid.length;
        const m = grid[0].length;
        const rotated = Array(m).fill().map(() => Array(n).fill(0));

        for (let i = 0; i < n; i++) {
            for (let j = 0; j < m; j++) {
                rotated[j][n - 1 - i] = grid[i][j];
                }
            }
        return rotated;
    }

    const rotated = rotate90(grid);
    return Math.min(solve(grid), solve(rotated));
};
